Library/Homebrew/utils/curl.rb

```diff
@@ -231,6 +232,52 @@ def curl_check_http_content(url, user_agents: [:default], check_content: false,
       "The URL #{url} may be able to use HTTPS rather than HTTP. Please verify it in a browser."
     end
 
+    # Separates cURL response text into headers and body. Headers (if any) are
+    # split into an array, where each member is the headers string for an
+    # individual response.
+    # @param response_text [String] The response text from `curl` containing headers, body, or both.
+    # @return [Array<Array<String>, String, nil>] An array containing an array of headers and the body content.
+    def response_headers_and_body(response_text)
+      header_text, _, body_text = response_text.rpartition("\r\n\r\n")
+      [header_text.split("\r\n\r\n"), body_text]
+    end
+
+    # Extracts the status code from the final response and the URL from the last
+    # `location` header encountered (if any).
+    # @param headers [Array<String>, String]
+    # @return [Array<String, nil>]
+    def response_status_code_and_location(headers, url: nil, absolutize: false)
+      # Separate the headers into individual responses if we're given a string
+      headers, = response_headers_and_body(headers) if headers.is_a?(String)
+
+      # Identify the last status code and location
+      status_code, final_location = nil
+      headers.each do |response_headers|
+        status_code = response_headers[%r{HTTP/.* (\d+)}, 1]
+        location = response_headers[/^Location:\s*(.*)$/i, 1]&.chomp
+        next unless location
+
+        final_location = if absolutize
+          if location.start_with?("//")
+            uri = URI(url)
+            "#{uri.scheme}:#{location}"
+          elsif location.start_with?("/")
+            uri = URI(url)
+            "#{uri.scheme}://#{uri.host}#{location}"
+          elsif location.start_with?("./")
+            uri = URI(url)
+            "#{uri.scheme}://#{uri.host}#{Pathname(uri.path).dirname/location}"
```

Comment:
I think `dirname` is actually wrong when the path ends with a slash, e.g.

```
http://example.org/subdir/ --> ./file1 --> http://example.org/file1
``` 

instead of 

```
http://example.org/subdir/ --> ./file1 --> http://example.org/subdir/file1
``` 

since `Pathname("subdir/").dirname` and `Pathname("subdir").dirname` are the same.

Reply:
Nice catch. I simply copied over the existing logic from  [`CurlDownloadStrategy#resolve_url_basename_time_file_size`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L425-L441), so any preexisting issues remain.

I'll start working on some tests to exercise this, as I hadn't yet tested the existing code that makes URLs absolute.

Reply:
I think in the existing code we only use it as a fallback for getting the file name so this bug would never manifest itself.

Reply:
With tests in place for these relative URL cases, I initially created a fix for the `./` case but wasn't quite satisfied with it. I tested `URI#join` and it correctly handles all of these cases, so I used it to replace the absolute URL code that I copied from [`CurlDownloadStrategy#resolve_url_basename_time_file_size`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L425-L441).

The end result is a simple one-liner that uses `URI#join`. I imagine `URI#join` is more correct/reliable than the code from `CurlDownloadStrategy`, so it seems like a better choice overall.